logstash-core/lib/logstash/agent.rb

```diff
@@ -10,80 +10,125 @@
   rescue => e
     @logger.fatal I18n.t("oops", :error => e)
     @logger.fatal e.backtrace if @logger.debug? || $DEBUGLIST.include?("stacktrace")
     return 1
-  ensure
-    Stud::untrap("INT", sigint_id) unless sigint_id.nil?
-    Stud::untrap("TERM", sigterm_id) unless sigterm_id.nil?
-  end # def execute
+  end
 
+  private
   def add_pipeline(pipeline_id, config_str, settings = {})
     @pipelines[pipeline_id] = LogStash::Pipeline.new(config_str, settings)
   end
 
-  private
-  # Emit a warning message.
-  def warn(message)
-    # For now, all warnings are fatal.
-    raise LogStash::ConfigurationError, message
-  end # def warn
+  def start_pipeline(id)
+    return unless @pipelines[id]
+    @logger.info("starting pipeline", :id => id)
+    @pipeline_threads[id] = Thread.new do
+      LogStash::Util.set_thread_name("pipeline.#{id}")
+      @pipelines[id].run
+    end
+  end
 
-  # Emit a failure message and abort.
-  def fail(message)
-    raise LogStash::ConfigurationError, message
-  end # def fail
+  def stop_pipeline(id)
+    return unless @pipelines[id]
+    @logger.warn("stopping pipeline", :id => id)
+    @pipelines[id].shutdown do
+      LogStash::ShutdownController.start(@pipelines[id], @pipeline_threads[id])
+    end
+    @pipeline_threads[id].join
+  end
 
   def shutdown_pipelines
-    @pipelines.each do |_, pipeline|
-      pipeline.shutdown do
-        ::LogStash::ShutdownController.start(pipeline)
-      end
-    end
+    @pipelines.each { |id, _| stop_pipeline(id) }
   end
 
-  def trap_sigterm
-    Stud::trap("TERM") do
-      @logger.warn(I18n.t("logstash.agent.sigterm"))
-      shutdown_pipelines
-    end
+  def running_pipelines?
+    @pipeline_threads.select {|_, pipeline| pipeline.alive? }.any?
   end
 
-  def trap_sigint
-    Stud::trap("INT") do
-      if @interrupted_once
-        @logger.fatal(I18n.t("logstash.agent.forced_sigint"))
-        exit
-      else
-        @logger.warn(I18n.t("logstash.agent.sigint"))
-        Thread.new(@logger) {|logger| sleep 5; logger.warn(I18n.t("logstash.agent.slow_shutdown")) }
-        @interrupted_once = true
-        shutdown_pipelines
-      end
-    end
+  # Override the methods below if you're implementing your own agent
+  def upgrade_state(new_state)
+    stop_pipeline("base")
+    add_pipeline("base", new_state, @pipeline_settings)
+  rescue => e
+    @logger.error("failed to update state", :new_state => new_state, :message => e.message, :backtrace => e.backtrace)
+    @logger.warn("reverting to previous state", :state => @state)
+    add_pipeline("base", @state, @pipeline_settings) unless clean_state?
+    @state
+  else
+    @state = new_state
+  ensure
+    start_pipeline("base") unless clean_state?
+  end
+
+  def fetch_state
+    @config_loader.format_config(@config_path, @config_string)
   end
+
+  def valid_state?(new_state)
+    new_state.is_a?(String)
+  end
+
+  def new_state?(old_state, new_state)
+    old_state != new_state
+  end
+
+  def clean_state
```

Comment:
different agents will have different notions of clean_state, for the default agent is "", but for another can be a `{"version" => 0}` hashmap for example